sys/net/network_layer/ng_ndp/ng_ndp.c

```diff
@@ -0,0 +1,950 @@
+
+            _send_nbr_sol(nc_entry->iface, &nc_entry->ipv6_addr, &dst);
+
+            mutex_lock(&ipv6_iface->mutex);
+            vtimer_set_msg(&nc_entry->nbr_sol_timer,
+                           ipv6_iface->retrans_timer, ng_ipv6_getpid(),
+                           NG_NDP_MSG_NBR_SOL_RETRANS, nc_entry);
+            mutex_unlock(&ipv6_iface->mutex);
+        }
+    }
+    else if (nc_entry->probes_remaining <= 1) {
+        ng_pktqueue_node_t *queue_node;
+
+        /* No need to call ng_ipv6_nc_remove() we know already were the
+         * entry is */
+
+        DEBUG("ndp: Remove nc entry %s for interface %" PRIkernel_pid "\n",
+              ng_ipv6_addr_to_str(addr_str, nc_entry->ipv6_addr, sizeof(addr_str)),
+              nc_entry->iface);
+
+        while ((queue_node = ng_pktqueue_remove_head(&nc_entry->pkts))) {
+            ng_pktbuf_release(queue_node->data);
+        }
+
+        ng_ipv6_addr_set_unspecified(&(nc_entry->ipv6_addr));
+        nc_entry->iface = KERNEL_PID_UNDEF;
+        nc_entry->flags = 0;
+        nc_entry->probes_remaining = 0;
+    }
+}
+
+void ng_ndp_state_timeout(ng_ipv6_nc_t *nc_entry)
+{
+    switch (ng_ipv6_nc_get_state(nc_entry)) {
+        case NG_IPV6_NC_STATE_REACHABLE:
+            _set_state(nc_entry, NG_IPV6_NC_STATE_STALE);
+            break;
+
+        case NG_IPV6_NC_STATE_DELAY:
+            _set_state(nc_entry, NG_IPV6_NC_STATE_PROBE);
+            break;
+
+        default:
+            break;
+    }
+}
+
+void ng_ndp_netif_add(ng_ipv6_netif_t *iface)
+{
+    uint32_t reach_time = _rand(NG_NDP_MIN_RAND, NG_NDP_MAX_RAND);
+
+    /* set default values */
+    mutex_lock(&iface->mutex);
+    iface->reach_time_base = NG_NDP_REACH_TIME;
+    reach_time = (reach_time * iface->reach_time_base) / 10;
+    iface->reach_time = timex_set(0, reach_time);
+    timex_normalize(&iface->reach_time);
+    iface->retrans_timer = timex_set(0, NG_NDP_RETRANS_TIMER);
+    timex_normalize(&iface->retrans_timer);
+    mutex_unlock(&iface->mutex);
+}
+
+void ng_ndp_netif_remove(ng_ipv6_netif_t *iface)
+{
+    /* TODO */
+}
+
+static ng_ipv6_addr_t *_default_router(void)
+{
+    ng_ipv6_nc_t *router = ng_ipv6_nc_get_next_router(NULL);
+
+    /* first look if there is any reachable router */
+    while (router != NULL) {
+        if ((ng_ipv6_nc_get_state(router) != NG_IPV6_NC_STATE_INCOMPLETE) &&
+            (ng_ipv6_nc_get_state(router) != NG_IPV6_NC_STATE_UNREACHABLE)) {
+            _last_router = NULL;
+
+            return &router->ipv6_addr;
+        }
+
+        router = ng_ipv6_nc_get_next_router(router);
+    }
+
+    /* else take the first one, but keep round-robin in further selections */
+    router = ng_ipv6_nc_get_next_router(_last_router);
+
+    if (router == NULL) {   /* end of router list or there is none => wrap around */
+        router = ng_ipv6_nc_get_next_router(router);
+
+        if (router == NULL) {   /* still nothing found => no router in list */
+            return NULL;
```

Comment:
Just for this one use-case? Seems inefficient to me... Especially since the code would basically look the same, just with an extra `&& !ng_ipv6_nc_is_empty()` in l320, if I see this correctly.